### DISCUSSION TITLE
SourceGenerators & partial methods

### DISCUSSION THREAD

**Discussion by DoctorVanGogh** [dotnet/roslyn#69387] [Q&A] [answered]
Writing a source generator I've come across a situation with partial methods where I am not sure if my scenario is just ~~unsupported~~ "iffy", or I'm doing something wrong.

Imagine a source generator which outputs source that contains partial method stubs:

`Foo.generated.cs`:
```cs
/// <autogenerated />
public partial class Foo {
  public partial object Bar(int wiz, string bang);
}
```
Since that is partial method with both a visibility modifier _and_ a return type (either would suffice) it _must_ have an implementation part. (See https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods#partial-methods).

Now imagine a second manually written source file containing said implementation:
`Foo.cs`
```cs
public partial class Foo {
  public partial object Bar(int wiz, string bang) {
     return $"{wiz}-{bang}";
  }
}
```
If both files were part of a regular project this would compile fine. If the manually written implementation was missing, you would get a nice compiler error for that.

---

Now if the _defining_ declaration comes out of a source generator the compiler may error with
`No defining declaration found for implementing declaration of partial method 'Foo.Bar(int, string)'`

I've specifically had those errors when using a non incremental `ISourceGenerator` - presumably since the non incremental generator seems to run _after_ all other compilations.
Using `IIncrementalGenerator` I've been able to sidestep this (mostly*) since the generated source seems to be created early enough for roslyn to include it in the analysis of the manually crafted file. 

I'm _assuming_ that combining my value providers with the `CompilationProvider` would once again put an arrow in my knee here.

Now my question is: Is this intentional? Shouldn't roslyn only give diagnostic errors for potentially multi-file issues **once it has all files** ?

**Comment by CyrusNajmabadi** [comment, marked as answer]
> Shouldn't roslyn only give diagnostic errors for potentially multi-file issues once it has all files ?

Yes. 

If you are seeing otherwise, it is a bug. We'd need a repro case demonstrating the issue so someone could investigate. Thanks!

**Reply by jasonmalinowski** [reply]
Indeed, there should be no change of behavior whether you're using ISourceGenerator or IIncrementalGenerator.

**Comment by DoctorVanGogh** [comment]
Okay, that's rather good news (and bad). 😅 

I was having significant issues getting the source generator to work in all build situations due to FileNotFound exceptions with (transitive) assembly dependencies. The missing definition errors _may_ just have coincided with those initialization errors (and at those times been correct from a compiler perspective).
I think I eliminated all my initialization issues at the same time as switching to the incremental generator. So it might just have been that (and me thinking  "Of _course_ it works now: The incremental generator creates the source as soon as the value provider chain is satisfied so the generated code exists _before_ any actual build")

I'll give it another look and maybe put in an issue or two regarding dependencies (and any other weird behavior I've seen).
